File: src/main/java/com/hazza/LeetCodeByJava/No40CombinationSumII.java
package com.hazza.LeetCodeByJava;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * https://leetcode.com/problems/combination-sum-ii/#/description
 */
public class No40CombinationSumII {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        if (candidates == null || candidates.length == 0 || target <= 0) return res;
        Arrays.sort(candidates);
        getConmbination(candidates, 0, candidates.length, target, res, new ArrayList<>());

        return res;
    }

    private void getConmbination(int[] candidates, int start, int end, int target,
                                 List<List<Integer>> res, List<Integer> path) {
        if (target < 0) return;
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = start; i < end; i++) {
            if (i > start && candidates[i] == candidates[i - 1]) continue;
            if (target < candidates[i]) break;
            path.add(candidates[i]);
            getConmbination(candidates, i + 1, end, target - candidates[i], res, path);
            path.remove(path.size() - 1);
        }
    }
}

File: src/main/python/com/hazza/leetcode_by_python/SegmentOne/No40_Combination_Sum_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/5/10 20:50
# https://leetcode.com/problems/combination-sum-ii/
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        self.dfs(candidates, target, 0, [], res)

        return res

    def dfs(self, candidates, target, idx, path, res):
        if target < 0:
            return
        if target == 0:
            res.append(path)
            return
        for i in range(idx, len(candidates)):
            if i > idx and candidates[i] == candidates[i - 1]:
                continue
            if candidates[i] > target:
                break
            self.dfs(candidates, target - candidates[i], i + 1, path + [candidates[i]], res)


if __name__ == '__main__':
    solution = Solution()

    assert sorted(solution.combinationSum2([10, 1, 2, 7, 6, 1, 5], 8)) == sorted([[1, 7], [1, 2, 5], [2, 6], [1, 1, 6]])
    assert sorted(solution.combinationSum2([2, 5, 2, 1, 2], 5)) == sorted([[1, 2, 2], [5]])